File: frontend/src/features/ai/code-factory/types/codeFactory.ts
export type RiskTier = 'low' | 'standard' | 'high' | 'critical';
export type ContractStatus = 'draft' | 'active' | 'archived';
export type ReviewStatus = 'pending' | 'reviewing' | 'clean' | 'dirty' | 'stale';
export type EvidenceStatus = 'pending' | 'captured' | 'verified' | 'failed';
export type GapStatus = 'open' | 'in_progress' | 'case_added' | 'verified' | 'closed';
export type GapSeverity = 'low' | 'medium' | 'high' | 'critical';
export type RunStatus = 'pending' | 'preflight' | 'reviewing' | 'remediating' | 'verifying' | 'evidence_capture' | 'completed' | 'failed';

export interface RiskRule {
  tier: RiskTier;
  patterns: string[];
  required_checks: string[];
  evidence_required: boolean;
  min_reviewers: number;
}

export interface MergePolicy {
  [tier: string]: {
    auto_merge: boolean;
    require_approval: boolean;
    min_approvals: number;
  };
}

export interface RiskContract {
  id: string;
  account_id: string;
  repository_id: string | null;
  created_by_id: string | null;
  name: string;
  version: number;
  status: ContractStatus;
  risk_tiers: RiskRule[];
  merge_policy: MergePolicy;
  docs_drift_rules: Record<string, unknown>;
  evidence_requirements: Record<string, unknown>;
  remediation_config: Record<string, unknown>;
  preflight_config: Record<string, unknown>;
  metadata: Record<string, unknown>;
  activated_at: string | null;
  created_at: string;
  updated_at: string;
  repository?: { id: string; name: string; full_name: string };
  created_by?: { id: string; name: string; email: string };
}

export interface ReviewState {
  id: string;
  account_id: string;
  risk_contract_id: string;
  repository_id: string | null;
  pr_number: number;
  head_sha: string;
  status: ReviewStatus;
  risk_tier: RiskTier | null;
  required_checks: string[];
  completed_checks: string[];
  evidence_verified: boolean;
  all_checks_passed: boolean;
  review_findings_count: number;
  critical_findings_count: number;
  remediation_attempts: number;
  bot_threads_resolved: number;
  stale_reason: string | null;
  reviewed_at: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  risk_contract?: { id: string; name: string };
  evidence_manifests?: EvidenceManifest[];
}

export interface EvidenceManifest {
  id: string;
  review_state_id: string;
  manifest_type: string;
  status: EvidenceStatus;
  assertions: EvidenceAssertion[];
  artifacts: EvidenceArtifact[];
  verification_result: Record<string, unknown>;
  captured_at: string | null;
  verified_at: string | null;
  created_at: string;
}

export interface EvidenceAssertion {
  type: string;
  selector: string;
  expected: unknown;
  actual: unknown;
  passed: boolean | null;
}

export interface EvidenceArtifact {
  type: string;
  url: string;
  sha256: string;
  size_bytes: number;
  captured_at: string;
}

export interface HarnessGap {
  id: string;
  account_id: string;
  risk_contract_id: string | null;
  incident_source: string;
  incident_id: string;
  description: string;
  status: GapStatus;
  severity: GapSeverity;
  test_case_added: boolean;
  test_case_reference: string | null;
  sla_deadline: string | null;
  sla_met: boolean | null;
  resolution_notes: string | null;
  resolved_at: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface PreflightResult {
  passed: boolean;
  risk_tier: RiskTier | null;
  required_checks: string[];
  evidence_required: boolean;
  review_state_id: string | null;
  reason: string | null;
}

export interface CodeFactoryMetrics {
  total_runs: number;
  auto_fix_rate: number;
  preflight_catch_rate: number;
  sla_compliance: number;
}

export interface HarnessGapMetrics {
  total: number;
  open: number;
  in_progress: number;
  closed: number;
  sla_compliance_rate: number;
  by_severity: Record<string, number>;
}

export interface SlaCompliance {
  total_open: number;
  past_sla_count: number;
  past_sla_gaps: Array<{
    id: string;
    incident_id: string;
    severity: string;
    sla_deadline: string;
    hours_overdue: number;
  }>;
}

export interface CodeFactoryWebSocketEvent {
  event: string;
  payload: Record<string, unknown>;
  timestamp: string;
}

File: server/spec/factories/ai/code_factory.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :ai_code_factory_risk_contract, class: "Ai::CodeFactory::RiskContract" do
    account
    sequence(:name) { |n| "Risk Contract #{n}" }
    status { "draft" }
    version { 1 }
    risk_tiers do
      [
        {
          "tier" => "low",
          "patterns" => ["docs/**", "*.md"],
          "required_checks" => ["lint"],
          "evidence_required" => false,
          "min_reviewers" => 0
        },
        {
          "tier" => "standard",
          "patterns" => ["app/**", "lib/**"],
          "required_checks" => ["lint", "tests"],
          "evidence_required" => false,
          "min_reviewers" => 1
        },
        {
          "tier" => "high",
          "patterns" => ["app/models/**", "db/migrate/**"],
          "required_checks" => ["lint", "tests", "security_scan"],
          "evidence_required" => true,
          "min_reviewers" => 2
        },
        {
          "tier" => "critical",
          "patterns" => ["config/credentials/**", "app/services/payments/**"],
          "required_checks" => ["lint", "tests", "security_scan", "manual_review"],
          "evidence_required" => true,
          "min_reviewers" => 3
        }
      ]
    end

    trait :active do
      status { "active" }
      activated_at { Time.current }
    end

    trait :archived do
      status { "archived" }
    end

    trait :with_repository do
      association :repository, factory: :devops_git_repository
    end
  end

  factory :ai_code_factory_review_state, class: "Ai::CodeFactory::ReviewState" do
    account
    association :risk_contract, factory: :ai_code_factory_risk_contract
    pr_number { rand(1..999) }
    head_sha { SecureRandom.hex(20) }
    status { "pending" }

    trait :reviewing do
      status { "reviewing" }
    end

    trait :clean do
      status { "clean" }
      all_checks_passed { true }
      reviewed_at { Time.current }
    end

    trait :dirty do
      status { "dirty" }
      all_checks_passed { false }
      review_findings_count { 3 }
      critical_findings_count { 1 }
      reviewed_at { Time.current }
    end

    trait :stale do
      status { "stale" }
      stale_reason { "New push detected" }
    end

    trait :with_repository do
      association :repository, factory: :devops_git_repository
    end
  end

  factory :ai_code_factory_evidence_manifest, class: "Ai::CodeFactory::EvidenceManifest" do
    account
    association :review_state, factory: :ai_code_factory_review_state
    manifest_type { "browser_test" }
    status { "pending" }

    trait :captured do
      status { "captured" }
      captured_at { Time.current }
      artifacts do
        [{ "type" => "screenshot", "url" => "https://example.com/screenshot.png", "sha256" => SecureRandom.hex(32), "size_bytes" => 1024 }]
      end
      assertions do
        [{ "type" => "element_exists", "selector" => "#main", "expected" => true, "actual" => true, "passed" => true }]
      end
    end

    trait :verified do
      status { "verified" }
      captured_at { 1.hour.ago }
      verified_at { Time.current }
      verification_result { { "passed" => true, "checks" => 3, "failures" => 0 } }
    end

    trait :failed do
      status { "failed" }
      captured_at { 1.hour.ago }
      verified_at { Time.current }
      verification_result { { "passed" => false, "checks" => 3, "failures" => 1 } }
    end
  end

  factory :ai_code_factory_harness_gap, class: "Ai::CodeFactory::HarnessGap" do
    account
    incident_source { "review_finding" }
    sequence(:incident_id) { |n| "INC-#{n}" }
    description { "Missing test coverage for authentication edge case" }
    status { "open" }
    severity { "medium" }

    trait :with_sla do
      sla_deadline { 72.hours.from_now }
    end

    trait :past_sla do
      sla_deadline { 24.hours.ago }
    end

    trait :with_test_case do
      status { "case_added" }
      test_case_added { true }
      test_case_reference { "spec/models/user_spec.rb:42" }
    end

    trait :closed do
      status { "closed" }
      resolved_at { Time.current }
      resolution_notes { "Test case added and verified" }
    end

    trait :critical do
      severity { "critical" }
    end
  end
end
